proposals/0000-custom-type-errors.rst

```diff
@@ -0,0 +1,188 @@
+
+    (==) :: CustomErrors [ a :~/: b :=>: Text "== is applied to arguments of different types"
+                         , Undischarged (Eq a) :=>: ShowType a :<>: Text " does not implement equality."]
+         => a -> b -> Bool
+
+You might have noticed that I have emphasized "at that point in the solving process" when describing ``IfApartUnsafe``. The reason is that we cannot defer indefinitely every apartness check, so sometimes we need to take the ``yes`` branch and continue. Thus, the order in which constraints are solved becomes quite important. I propose to change the default from "don't care" to an order in which constraints coming from arguments to a function are given priority with respect to those coming from the application itself.
+
+Updates to the solver
+~~~~~~~~~~~~~~~~~~~~~
+Inside GHC solver each constraint would come with an optional attached message and a priority number. Whenever a new item has to be taken out of the work list, the ones with highest priority should come before. If more than one constraint have the same priority, then the choice is done as now -- first canonicalization, then interaction, and so on.
+
+Whenever the solver finds a ``IfUndischarged c msg``, it emits the constraint ``c`` with the additional message attached. This message has no influence on the rest of the solving process, but if it ends with errors and the constraint should be reported, the message is used. Indeed, we could see its definition as (note that ``UndecidableSuperClasses`` is required) ::
+
+    class c => IfUndischarged c msg
+
+Priorities and messages are inherited by constraints stemming from another one by solving. For example, if we have ``IfUndischarged (Eq [a]) msg``, we would end up with a constraint ``Eq a`` where the message ``msg`` is kept.
+
+Finally, the solver has to be updated with the following new rules for inequality checks ::
+
+    [G] IfApartUnsafe a b no yes ==> yes
+    [W] IfApartUnsafe a b no yes
+          | if a and b are apart ==> no
+          | otherwise            ==> yes
+
+Indeed, we could see `IfApartUnsafe` as defined whenever compilation is successful as (again, using ``UndecidableSuperClasses``) ::
+
+    class yes => IfApartUnsafe a b no yes
+
+
+Effect and Interactions
+-----------------------
+In principle, there should be no effect for already-existing code. Right now people do not assume any specific ordering on the constraint solver, which means that we can change it as explained here.
+
+If implemented as-is, using ``(==)`` with the aforementioned type signature would impose a performance cost. Whereas before we just needed a ``Eq`` dictionary and a type application, now we need two of them and a complex proof of the big constraint. Luckily, this is easy to fix by using an ``INLINE`` pragma ::
```

Comment:
Are you worried about compile-time performance or runtime performance? It looks like the latter. But I don't agree that this would be a problem. Have you measured this cost? The difference between a proof of `Eq a` and of `IfUndischarged (Eq a) ...` could be the same at runtime. (That is, there is a cast necessary in the latter, but casts are erased during compilation.)